Neveikia /datasets

```python
Internal Server Error: /datasets/
Traceback (most recent call last):
  File "django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "django/core/handlers/base.py", line 181, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "django/views/generic/base.py", line 70, in view
    return self.dispatch(request, *args, **kwargs)
  File "django/views/generic/base.py", line 98, in dispatch
    return handler(request, *args, **kwargs)
  File "haystack/generic_views.py", line 130, in get
    return self.form_valid(form)
  File "haystack/generic_views.py", line 77, in form_valid
    context = self.get_context_data(
  File "vitrina/datasets/views.py", line 150, in get_context_data
    date_facets = context['facets']['dates']
KeyError: 'dates'
```

Klaida pasireiškia neprisijungusiems naudotojams.

Būtina parašyti testą, kuri atkartoja klaidą. Kad ateityje tokio pobūdžio klaidų nebūtų.

`rebuild_index` yra paleistas, jis daromas po kiekvieno atnaujinimo. Jei būtų indekso problema, tai tikriausiai neveiktu ir su prisijungusiu naudotoju.

Po atnaujinimo gaunu naują klaidą:

```python
Traceback (most recent call last):
  File "haystack/backends/elasticsearch_backend.py", line 585, in search
    raw_results = self.conn.search(
  File "elasticsearch/client/utils.py", line 139, in _wrapped
    return func(*args, params=params, headers=headers, **kwargs)
  File "elasticsearch/client/__init__.py", line 1479, in search
    return self.transport.perform_request(
  File "elasticsearch/transport.py", line 345, in perform_request
    status, headers_response, data = connection.perform_request(
  File "elasticsearch/connection/http_urllib3.py", line 256, in perform_request
    self._raise_error(response.status, raw_data)
  File "elasticsearch/connection/base.py", line 287, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(
elasticsearch.exceptions.RequestError:
    RequestError(400, 'search_phase_execution_exception', 'Failed to parse query [django_id:(
        "2" OR
        "3" OR
        "4" OR
        [...]  (1525 items ommited)
        "2350" OR
        "2355" OR
        "2356"
    )]')
Internal Server Error: /datasets/
Traceback (most recent call last):
  File "django/core/handlers/exception.py", line 47, in inner
    response = get_response(request)
  File "django/template/base.py", line 905, in render_annotated
    return self.render(context)
  File "django/template/defaulttags.py", line 170, in render
    values = list(values)
  File "vitrina/helpers.py", line 99, in items
    facet = fields[self.name]
TypeError: 'NoneType' object is not subscriptable
```

ElasticSearch loguose matau tokias žinutes:

```
Caused by: org.apache.lucene.queryparser.classic.ParseException: Cannot parse:
    django_id:(
        "2" OR
        "3" OR
        "4" OR
        [...]  (1525 items ommited)
        "2350" OR
        "2355" OR
        "2356"
    )'
too many boolean clauses.
```

Jei bandoma į vieną ElasticSearch užklausą sudėti visų duomenų rinkinių ID į vieną didelę OR seką, tai anksčiau ar vėliau ir turėjo nustoti veikit.

--- a/tests/datasets/test_views.py
+++ b/tests/datasets/test_views.py
@@ -125,6 +125,34 @@ def test_dataset_list_view_anon_user_without_datasets(app: DjangoTestApp):
 
 
 @pytest.mark.haystack
+def test_dataset_list_view_all_shown_for_staff(app: DjangoTestApp):
+    org1 = OrganizationFactory()
+    org2 = OrganizationFactory()
+    DatasetFactory(organization=org1, is_public=False)
+    DatasetFactory(organization=org1)
+    DatasetFactory(organization=org2)
+    DatasetFactory(organization=org2, is_public=False)
+    user = UserFactory(is_staff=True)
+    app.set_user(user)
+    resp = app.get(reverse('dataset-list'))
+    assert len(resp.context['object_list']) == 4
+
+
+@pytest.mark.haystack
+def test_dataset_list_view_public_shown_for_regular_user(app: DjangoTestApp):
+    org1 = OrganizationFactory()
+    org2 = OrganizationFactory()
+    DatasetFactory(organization=org1, is_public=False)
+    DatasetFactory(organization=org1)
+    DatasetFactory(organization=org2)
+    DatasetFactory(organization=org2, is_public=False)
+    user = UserFactory()
+    app.set_user(user)
+    resp = app.get(reverse('dataset-list'))
+    assert len(resp.context['object_list']) == 2
+
+
+@pytest.mark.haystack
 def test_org_dataset_url_is_hidden_for_anon_user(app: DjangoTestApp):
     resp = app.get(reverse('dataset-list'))
     assert not resp.html.find(id='org-dataset-url')

--- a/vitrina/datasets/models.py
+++ b/vitrina/datasets/models.py
@@ -15,7 +15,7 @@ from random import randrange
 
 from vitrina.structure.models import Model, Base, Property, Metadata
 from vitrina.users.models import User
-from vitrina.orgs.models import Organization
+from vitrina.orgs.models import Organization, Representative
 from vitrina.catalogs.models import Catalog, HarvestingJob
 from vitrina.classifiers.models import Category, Licence, Frequency
 from vitrina.datasets.managers import PublicDatasetManager
@@ -291,6 +291,12 @@ class Dataset(TranslatableModel):
 
     def get_members_url(self):
         return reverse('dataset-members', kwargs={'pk': self.pk})
+
+    def get_managers(self):
+        ct = ContentType.objects.get_for_model(Dataset)
+        return list(Representative.objects.filter(
+            content_type=ct, object_id=self.id
+        ).values_list('user_id', flat=True))
 
     @property
     def language_array(self):

--- a/vitrina/datasets/search_indexes.py
+++ b/vitrina/datasets/search_indexes.py
@@ -1,4 +1,4 @@
-from haystack.fields import CharField, IntegerField, MultiValueField, DateTimeField, EdgeNgramField
+from haystack.fields import CharField, IntegerField, MultiValueField, DateTimeField, EdgeNgramField, BooleanField
 from django.db import models
 
 from haystack import signals
@@ -28,6 +28,8 @@ class DatasetIndex(SearchIndex, Indexable):
     level = IntegerField(model_attr='get_level', faceted=True, null=True)
     type = MultiValueField(model_attr='public_types', faceted=True)
     type_order = IntegerField(model_attr='type_order')
+    is_public = BooleanField(model_attr='is_public', faceted=True, null=False)
+    managers = MultiValueField(model_attr='get_managers', faceted=True)
 
     def get_model(self):
         return Dataset

--- a/vitrina/datasets/services.py
+++ b/vitrina/datasets/services.py
@@ -6,10 +6,9 @@ from django.core.handlers.wsgi import HttpRequest
 
 from django.db.models import Q, Sum
 from django.db.models.functions import ExtractQuarter, ExtractWeek
+from haystack.backends import SQ
 
-from vitrina.datasets.models import Dataset
 from vitrina.helpers import get_filter_url
-from vitrina.orgs.models import Representative, Organization
 from vitrina.projects.models import Project
 from vitrina.requests.models import Request, RequestObject
 
@@ -156,56 +155,14 @@ def get_frequency_and_format(duration):
     return frequency, ff
 
 
-def get_public_dataset_id_list():
-    public_datasets = Dataset.objects.filter(is_public=True)
-    public_dataset_id_list = [dataset.id for dataset in public_datasets]
-    return public_dataset_id_list
-
-
-def filter_datasets_for_user(user, datasets):
-    coordinator_orgs = [rep.object_id for rep in
-                        user.representative_set.filter(content_type=ContentType.objects.get_for_model(Organization))]
-    public_dataset_id_list = get_public_dataset_id_list()
-    coordinated_datasets = Dataset.objects.filter(organization_id__in=coordinator_orgs)
-    dataset_id_list = public_dataset_id_list + [dataset.id for dataset in coordinated_datasets]
-    datasets = datasets.filter(django_id__in=dataset_id_list)
-    return datasets
-
-
 def get_datasets_for_user(user, datasets):
     if user.is_authenticated:
         if not (user.is_staff or user.is_superuser):
-            return filter_datasets_for_user(user, datasets)
+            if user.representative_set:
+                return datasets.filter(SQ(is_public='true') | SQ(managers__contains=user.id))
+            else:
+                return datasets.filter(is_public='true')
         else:
             return datasets
     else:
-        id_list = get_public_dataset_id_list()
-        datasets = datasets.filter(django_id__in=id_list)
-        return datasets
-
-
-def get_all_not_deleted_orgs():
-    orgs = Organization.objects.filter(deleted__isnull=True, deleted_on__isnull=True)
-    return orgs
-
-
-def filter_orgs_for_user(user):
-    non_deleted_orgs = get_all_not_deleted_orgs()
-    public_orgs = [org.id for org in non_deleted_orgs.filter(is_public=True)]
-    coordinator_orgs = [org.object_id for org in
-                        user.representative_set.filter(role=Representative.COORDINATOR)]
-    manager_orgs = [org.object_id for org in
-                    user.representative_set.filter(role=Representative.MANAGER)]
-    org_ids = coordinator_orgs + public_orgs + manager_orgs
-    orgs = Organization.objects.filter(id__in=org_ids)
-    return orgs
-
-
-def get_orgs_for_user(user):
-    if user.is_authenticated:
-        if not (user.is_staff or user.is_superuser):
-            return filter_orgs_for_user(user)
-        else:
-            return get_all_not_deleted_orgs()
-    else:
-        return Organization.public.all()
+        return datasets.filter(is_public='true')

--- a/vitrina/datasets/views.py
+++ b/vitrina/datasets/views.py
@@ -322,6 +322,7 @@ class DatasetDetailView(
     HistoryMixin,
     DatasetStructureMixin,
     PlanMixin,
+    PermissionRequiredMixin,
     DetailView
 ):
     model = Dataset
@@ -330,6 +331,13 @@ class DatasetDetailView(
     detail_url_name = 'dataset-detail'
     history_url_name = 'dataset-history'
     plan_url_name = 'dataset-plans'
+
+    def has_permission(self):
+        dataset = get_object_or_404(Dataset, id=self.kwargs['pk'])
+        if dataset.is_public:
+            return True
+        else:
+            return has_perm(self.request.user, Action.VIEW, dataset)
 
     def get_context_data(self, **kwargs):
         context_data = super().get_context_data(**kwargs)
@@ -787,6 +795,7 @@ class CreateMemberView(
                 representative=self.object
             )
 
+        self.dataset.save()
         return HttpResponseRedirect(self.get_success_url())
 
 
@@ -898,6 +907,7 @@ class UpdateMemberView(
             self.object.user.organization = self.dataset.organization
             self.object.user.save()
 
+        self.dataset.save()
         return HttpResponseRedirect(self.get_success_url())
 
 
@@ -920,6 +930,13 @@ class DeleteMemberView(
         return reverse('dataset-members', kwargs={
             'pk': self.kwargs.get('dataset_id'),
         })
+
+    def delete(self, request, *args, **kwargs):
+        super().delete((self, request, args, kwargs))
+        if self.object.content_type == ContentType.objects.get_for_model(Dataset):
+            dataset = Dataset.objects.get(id=self.object.object_id)
+            dataset.save()
+        return HttpResponseRedirect(self.get_success_url())
 
 
 class DatasetProjectsView(

--- a/vitrina/orgs/models.py
+++ b/vitrina/orgs/models.py
@@ -140,6 +140,8 @@ class Representative(models.Model):
     object_id = models.PositiveIntegerField()
     content_object = GenericForeignKey('content_type', 'object_id')
 
+    objects = models.Manager()
+
     class Meta:
         db_table = 'representative'
         unique_together = ['content_type', 'object_id', 'user']

--- a/vitrina/orgs/services.py
+++ b/vitrina/orgs/services.py
@@ -49,6 +49,7 @@ acl = {
     (Dataset, Action.HISTORY_VIEW): [Role.COORDINATOR, Role.MANAGER],
     (Dataset, Action.STRUCTURE): [Role.COORDINATOR, Role.MANAGER],
     (Dataset, Action.PLAN): [Role.COORDINATOR, Role.MANAGER],
+    (Dataset, Action.VIEW): [Role.COORDINATOR],
     (DatasetDistribution, Action.CREATE): [Role.COORDINATOR, Role.MANAGER],
     (DatasetDistribution, Action.UPDATE): [Role.COORDINATOR, Role.MANAGER],
     (DatasetDistribution, Action.DELETE): [Role.COORDINATOR, Role.MANAGER],

--- a/vitrina/orgs/views.py
+++ b/vitrina/orgs/views.py
@@ -22,11 +22,9 @@ from reversion.views import RevisionMixin
 from vitrina import settings
 from vitrina.api.models import ApiKey
 from vitrina.datasets.models import Dataset
-from vitrina.datasets.services import get_orgs_for_user, get_all_not_deleted_orgs
 from vitrina.helpers import get_current_domain
-from vitrina.orgs.forms import RepresentativeUpdateForm, OrganizationPlanForm, OrganizationMergeForm
-from vitrina.orgs.forms import RepresentativeCreateForm, RepresentativeUpdateForm, \
-    PartnerRegisterForm
+from vitrina.orgs.forms import OrganizationPlanForm, OrganizationMergeForm
+from vitrina.orgs.forms import RepresentativeCreateForm, RepresentativeUpdateForm, PartnerRegisterForm
 from vitrina.orgs.models import Organization, Representative
 from vitrina.orgs.services import has_perm, Action
 from vitrina.plans.models import Plan
@@ -47,7 +45,7 @@ class OrganizationListView(ListView):
     def get_queryset(self):
         query = self.request.GET.get('q')
         jurisdiction = self.request.GET.get('jurisdiction')
-        orgs = get_orgs_for_user(self.request.user)
+        orgs = Organization.public.all()
 
         if query:
             orgs = orgs.filter(title__icontains=query)

